vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/grid/it/GridBenchmark.java

```diff
@@ -0,0 +1,179 @@
+    private static final TreeData<String> treeData = new TreeData<>();
+
+    static {
+        addTreeItems(treeData, null, 1000, 2);
+    }
+
+    private static void addTreeItems(TreeData<String> treeData, String parent, int count, int level) {
+        IntStream.range(0, count).forEach(index -> {
+            String child = parent != null ? parent + "-" + index : String.valueOf(index);
+            treeData.addItem(parent, child);
+            if (level > 0) {
+                addTreeItems(treeData, child, 3, level - 1);
+            }
+        });
+    }
+
+    @Override
+    public void setParameter(BeforeEvent event, @OptionalParameter String parameter) {
+        Location location = event.getLocation();
+        QueryParameters queryParameters = location.getQueryParameters();
+
+        Map<String, List<String>> parametersMap = queryParameters.getParameters();
+        if (!parametersMap.containsKey("variant") || !parametersMap.containsKey("metric")) {
+            add(new Text("Provide query parameters: variant and metric"));
+            return;
+        }
+        String metric = parametersMap.get("metric").get(0);
+        String variant = parametersMap.get("variant").get(0);
+
+        LoggerFactory.getLogger(GridBenchmark.class).warn("Sample: " + variant + "-" + metric);
+
+        switch (variant) {
+            case "simple":
+                grid = getGrid();
+                addColumns(grid, 5, false);
+                break;
+            case "multicolumn":
+                grid = getGrid();
+                addColumns(grid, 50, false);
+                break;
+            case "componentrenderers":
+                grid = getGrid();
+                addColumns(grid, 5, true);
+                break;
+            case "detailsopened":
+                grid = getGrid();
+                addColumns(grid, 5, false);
+                grid.setItemDetailsRenderer(new ComponentRenderer<>(item -> new Text(item.toString())));
+                items.forEach(item -> grid.setDetailsVisible(item, true));
+                break;
+            case "tree":
+                grid = getTreeGrid();
+                ((TreeGrid<String>) grid).addHierarchyColumn(i -> i);
+                addColumns(grid, 5, false);
+                break;
+            case "mixed":
+                grid = getTreeGrid();
+                ((TreeGrid<String>) grid).addHierarchyColumn(i -> i);
+                addColumns(grid, 50, true);
+                grid.setItemDetailsRenderer(new ComponentRenderer<>(item -> new Text(item.toString())));
+                treeData.getRootItems().forEach(item -> grid.setDetailsVisible(item, true));
+                break;
+            default:
+                break;
+        }
+
+        switch (metric) {
+            case "verticalscrollframetime":
+                add(grid);
+                whenRendered(grid).then(v -> grid.getElement().executeJs("window.measureScrollFrameTime(this, false)"));
```

Comment:
I'm not sure I understand the use of `PendingJavaScriptResult::then` here and other cases below.

You've used it for JavaScript methods that return `Promises`, but I don't think the `PendingJavaScriptResult::then` waits for promises to resolve. Based on the JavaDoc, the callback is simply invoked with the js-function's return value after the roundtrip.

Have I misunderstood the logic here?

Reply:
Seems to properly wait for the promise to resolve. Just tested with
```js
_window.startWhenRendered = (grid) => {
  return new Promise(resolve => {
    setTimeout(() => resolve(), 10000);
  });
};
```

And expandtime tests wait for 10 seconds before starting

Reply:
Sorry, my bad. I don't know what docs I was looking at previously, but now I can see the JavaDocs clearly mention that it can handle Promises.